core/src/main/java/com/linecorp/armeria/client/retry/AbstractRetryingClientBuilder.java

```diff
@@ -39,61 +39,60 @@
 public abstract class AbstractRetryingClientBuilder<O extends Response> {
 
     @Nullable
-    private final RetryRule retryRule;
+    private final RetryConfigBuilder<O> retryConfig;
 
     @Nullable
-    private final RetryRuleWithContent<O> retryRuleWithContent;
-
-    private int maxTotalAttempts = Flags.defaultMaxTotalAttempts();
-    private long responseTimeoutMillisForEachAttempt = Flags.defaultResponseTimeoutMillis();
+    private final RetryConfigMapping<O> mapping;
 
     /**
-     * Creates a new builder with the specified {@link RetryRule}.
+     * Creates a new builder with the specified {@link RetryConfig}.
      */
-    AbstractRetryingClientBuilder(RetryRule retryRule) {
-        this(requireNonNull(retryRule, "retryRule"), null);
+    AbstractRetryingClientBuilder(RetryConfig<O> retryConfig) {
+        this.retryConfig = requireNonNull(retryConfig, "retryConfig").toBuilder();
+        mapping = null;
     }
 
     /**
-     * Creates a new builder with the specified {@link RetryRuleWithContent}.
+     * Creates a new builder with the specified {@link RetryConfigMapping}.
      */
-    AbstractRetryingClientBuilder(RetryRuleWithContent<O> retryRuleWithContent) {
-        this(null, requireNonNull(retryRuleWithContent, "retryRuleWithContent"));
-    }
-
-    private AbstractRetryingClientBuilder(@Nullable RetryRule retryRule,
-                                          @Nullable RetryRuleWithContent<O> retryRuleWithContent) {
-        this.retryRule = retryRule;
-        this.retryRuleWithContent = retryRuleWithContent;
+    AbstractRetryingClientBuilder(RetryConfigMapping<O> mapping) {
+        this.mapping = requireNonNull(mapping, "mapping");
+        retryConfig = null;
     }
 
-    final RetryRule retryRule() {
-        checkState(retryRule != null, "retryRule is not set.");
-        return retryRule;
+    final RetryConfigMapping<O> mapping() {
+        if (mapping == null) {
+            final RetryConfig<O> config = retryConfig();
+            return (ctx, req) -> config;
```

Comment:
Could you add `assert config != null;` so that IDE does not complain?

Reply:
Done